SCClassLibrary/Common/Control/Server.sc

```diff
@@ -360,7 +360,7 @@ Server {
 		var n = options.maxLogins ? 1;
 
 		numControl = options.numControlBusChannels div: n;
-		numAudio = options.numAudioBusChannels div: n;
+		numAudio = options.numPrivateAudioBusChannels div: n;
```

Comment:
Oh, I ran out of audio busses all the time.  the structures I built using
crucial-library could consume quite a number of busses very quickly.
but I never noticed it trying to use a bus that didn't exist on the server.

For supercollider.js I've done a completely new implementation:

https://github.com/crucialfelix/supercolliderjs/blob/develop/src%2Fserver%2Finternals%2Fallocators.js
https://github.com/crucialfelix/supercolliderjs/blob/develop/src%2Fserver%2FServerState.js

its functional and uses immutable structures.

the server state is a single Immutable.
so if you are preparing a large structure and allocating many busses and
you run out (throws an error) then the operation fails, but the server
keeps its previous state intact.

a lot of issues with Busses or Nodes being out of sync disappeared. there
is one source of truth; there are no objects holding separate state.


On Thu, Feb 4, 2016 at 8:04 AM jamshark70 <notifications@github.com> wrote:

> In SCClassLibrary/Common/Control/Server.sc
> <https://github.com/supercollider/supercollider/pull/1845#discussion_r51836244>
> :
>
> > @@ -360,7 +360,7 @@ Server {
> >  		var n = options.maxLogins ? 1;
> >
> >  		numControl = options.numControlBusChannels div: n;
> > -		numAudio = options.numAudioBusChannels div: n;
> > +		numAudio = options.numPrivateAudioBusChannels div: n;
>
> Yes, seriously.
>
> Based on [1], it seems that the bug dates back more than 10 years --
> possibly all the way back to JMc's original implementation. [1] is my
> change to introduce ContiguousBlockAllocator, but I copied the
> (numAudioBusChannels, firstPrivateBus) mistake from the previous usage of
> PowerOfTwoAllocator!
>
> This means, nobody ever ran out of audio buses -- not in some 14-15 years.
>
> [1] 95065fd
> <https://github.com/supercollider/supercollider/commit/95065fd8f0a3bdd99c77200d8d928e15f5d3384d>
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/supercollider/supercollider/pull/1845/files#r51836244>
> .
>
-- 
https://twitter.com/crucialfelix
https://medium.com/@crucialfelix
http://www.mattermind.com/